controllers/api.template.js

```diff
@@ -22,7 +24,23 @@ router.get('/tpl/:id', async (req, res, next) => {
       return;
     }
 
-   // Increment template downloads
+    // Check no referrer
+    if(!req.headers.referer){        
+        res.writeHead(302, {
+          'Location': '/template/'+template.id+'/'+template.slug
+        });     
+        res.end();  
+    }      
+    // Check no-allowed-referrer
+    const referer_hostname= new URL(req.headers.referer).hostname;      
+    if(allowed_referrers.indexOf(referer_hostname)===-1){           
+        res.writeHead(302, {
+          'Location': '/template/'+template.id+'/'+template.slug
+        });     
+        res.end();  
+    }   
+      
```

Comment:
These checks should happen right at the top of this router, before the ID is checked and before any datastore interactions. An invalid referrer invalidates all the code in the whole router so there's no sense in resolving anything before this check.

Reply:
I'm doing the checks after queries becasue I need the slug in order to redirect the user. Another option may be at tamplate check if there's a slug and if not doing an extra redirect

ie: hotlink sends user to:  /templates/{id} 
in the template if there's no {{slug}} - redirect user to page with slug, still this will need a query too ..

Reply:
Can't you just redirect to path without /tpl? That will invoke the template layout which in turn will check the ID

Reply:
Sorry, confused myself there :)

Yeah, maybe we should just add on template to redirect user to slug, since it really adds no extra detail to the URL (as the ID is already unique).

Reply:
another option is sending user to /template/{{ID}} and via JS update the URL if the slug doesn't have an slug. 

Current Templates Views will be tracked using PDP views, so it's not a big deal

Reply:
won't work "router.get('/:id/:name')"  name is needed

Reply:
If i remove the name from the router when it's present its 404ing htrhrhr

Reply:
Can't you just add `?` so it becomes `:name?` thus becoming optional?

And no client-side redirects! :) We have a fully functional web server that we control - no need for hacks.

Reply:
I was not referring to JS redirect, I was pointing to a history update . Let me give it a try